### DISCUSSION TITLE
How to display MediaField in Custom Part

### DISCUSSION THREAD

**Discussion by TalMcMahon** [OrchardCMS/OrchardCore#8232] [Q&A] [answered]
Hello all, am trying to make a custom part that has a MediaField that I can use for a thumbnail to open a modal with a YoutubeField showing a youtube video. 
**I have made the following model:**
```
public class AslVideoPart : ContentPart
    {
        public string Title { get; set; }
        public string Description { get; set; }
        public string Label { get; set; }
        public MediaField Thumbnail {get;set; }
        public YoutubeField YoutubeLink { get; set; }
    }
```

How do I now create a Custom View (.cshtml) that has a MediaField that I can style?  
The media Field seems to render outside (before) the rendering of my CustomView. 
In my page content template I have this:
```
<h2>
    @T["Page Header"]
</h2>
@await DisplayAsync(@Model.Content)
<h3>
    @T["Page End"]
</h3>
```
When it renders, I see this:
```
<div class="col-lg-8 col-md-10 mx-auto">
<h2>
    Page Header
</h2>
    <div class="field field-type-mediafield field-name-asl-video-part-thumbnail">
        <span class="name">Asl Video Thumbnail</span>
                <img src="/media/VideoThumb.png" alt="VideoThumb.png">
    </div>
    <div class="field field-type-textfield field-name-asl-video-part-youtube-link">
                <iframe width="640" height="640" src="https://www.youtube.com/embed//F-y9JlYAcMo" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen=""></iframe>
    </div>
<h5>This is the Part</h5>
<h3>
    Page End
</h3>
</div>
```
As you can see, the img and iframe are rendered "magically" before the <h5>this is the part</h5> code
Any help would be greatly appreciated

Tal

**Comment by jtkech** [comment]
@TalMcMahon 

You need to define a shape template view that overrides the `MediaField.cshtml` of the `OC.Media` module. This in your current front end theme, or in a module that depends on `OC.Media` (a dependency in its manifest.cs file).

Then see https://docs.orchardcore.net/en/dev/docs/reference/modules/Templates/ to see all combinations with content type, part type / name, field type / name, and display type, that you can use for the view name

Just tried with an `Article` content type containing a media field (but not under a part) and with an `Article-MediaField.cshtml` override. In your case i think that you can use `AslVideoPart-MediaField.cshtml`

Then see what is done by default in the `MediaField.cshtml` of the `OC.Media` module, and that you can copy / paste before doing your customization.

**Reply by TalMcMahon** [reply]
Hi I see what you are saying. Do I understand this correctly, that by exactly naming the override with [PartType]__[FieldName] that it will pull inside of the part and I can place it within my template?

A follow up question, in my AslPartViewModel, do I have a property for MediaField or do I have to use DisplayMediaFieldViewModel like is in the OC.Media code?

Once I get this all figured out, I think I will for a PR for the documentation. The templating docs explain how it works, but I was unable to take those words and make the leap to what the implications were.

Thanks

Tal

**Comment by jtkech** [comment, marked as answer]
@TalMcMahon 

So, i assume you already have a view to render your custom part, then the most common way is to add another view, e.g `AslVideoPart-MediaField.cshtml`, to override the media field, and where the model will be `DisplayMediaFieldViewModel`, as in the original view. You can use a `@model ` directive to use a strongly typed model, or not and use it as a dynamic object.

You can also do all the job in you part view, here some example how you can access model properties, but as said in my tests the media field is not hold by a content part. Note: In fact a content type always has a special part whose name is the same as the content type itself, in my case `Article`, and this part is used to hold the fields that are directly under the type, not hold by a part

    @await DisplayAsync(Model.Content)

    <div>Testing</div>

    @* Through 'Model.Content' that is a dynamic Shape a zone shape holding child shapes *@
    @* Here, 'HtmlBodyPart' is a child shape on which you can call 'DisplayAsync()' *@
    @* Or just use some dynamic properties, but without using shape rendering *@
    <div>@Model.Content.HtmlBodyPart.Html</div>

    @* Another way to access a child shape whose name contains a '- '*@
    <div>@Model.Content.Named("Article-Image").Paths[0]</div>

    @* You can also use 'Model.ContentItem.Content' that is a JObject, as saved in the database *@
    @* Here to show the full json object and then see all its properties *@
    @Model.ContentItem.Content

    @* Here to render some properties *@
    @Model.ContentItem.Content["Article"]["Image"]["Paths"][0]

    @{ 
        // Finally, you can use on the 'ContentItem' some helpers
        // E.g. the 'As<>()' to retrieve a strongly typed 'HtmlBodyPart', here you need a reference on the 'OC.Html' module
        var bodyPart = ((ContentItem)Model.ContentItem).As<HtmlBodyPart>();
        var html = bodyPart.Html;
    }

**Reply by TalMcMahon** [reply]
Hi, I had already worked out the reading of the properties part similar to your first example.  This will work for me as I am compositing the image and the Youtube link into a single part anyway.

I was never able to get the @await DisplayAsync(Model.Content) to render properly, I am assuming I missed something in the setup.

Thanks for your time.

Tal
